### PR TITLE
Fix hear

### PR DESCRIPTION
**Pull request by ChrisCinelli**
More performant `hears` function.
It only listen to one event for each hear.
Precompile regEx.
Add the option to have a match function instead of a string
With debug on, we show the rule that matched.

### CODE REVIEW

**Comment by anonrig** [comment]
Please follow the JSDOC for documentation.

File: lib/CoreBot.js [hunk 1]
```diff
@@ -610,31 +610,48 @@ function Botkit(configuration) {
...
-        if (typeof(keywords) == 'string') {
-            keywords = [keywords];
-        }
-        if (typeof(events) == 'string') {
-            events = events.split(/\,/g);
+        var matchFn;
+        var theRule;
+
+        // Compile
+        if (typeof(keywords) == 'function') {
```
Comment (anonrig) [inline lib/CoreBot.js]: I believe this should be `===`

File: lib/CoreBot.js [hunk 2]
```diff
@@ -610,31 +610,48 @@ function Botkit(configuration) {
...
+            matchFn = keywords;
+            // Store first line of the fuction
+            theRule = (function strText(functionText){return functionText.substr(0, functionText.indexOf("\n"));})(keywords)
```
Comment (anonrig) [inline lib/CoreBot.js]: There is a 120 length limit per line. Will appreciate if you don't make it a single line.

**Comment by anonrig** [comment]
Test for this function is missing. Please include your tests.

File: lib/CoreBot.js (hunk 1 above)
Comment (ChrisCinelli) [inline lib/CoreBot.js]: both `==` and `===` works. It is a string comparison and  `==` was in the previous code.

File: lib/CoreBot.js (hunk 2 above)
Comment (ChrisCinelli) [inline lib/CoreBot.js]: Fixed

**Comment by ChrisCinelli** [comment]
No function in `CoreBot.js` is tested. If you guys set up tests, I will enhance the test of `hears` to cover the case for `function` as `keywords`.

**Comment by benbrown** [comment]
I have done what is probably a pretty similar refactor of the way hears works, in addition to implementing a new middleware system which could be used to extend hears to receive functions. 

Take a look:
https://github.com/howdyai/botkit/tree/experimental

**Pull request closed without merging**